File: backend/api/src/main/java/com/flowboard/api/task/TaskService.java
package com.flowboard.api.task;

import com.flowboard.api.project.ProjectRepository;
import com.flowboard.api.project.ResourceNotFoundException;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class TaskService {

    private final TaskRepository taskRepo;
    private final ProjectRepository projectRepo;

    public TaskService(TaskRepository taskRepo, ProjectRepository projectRepo) {
        this.taskRepo = taskRepo;
        this.projectRepo = projectRepo;
    }

    public List<TaskDto> findByProject(Long projectId) {
        if (!projectRepo.existsById(projectId)) {
            throw new ResourceNotFoundException("Project not found with id: " + projectId);
        }
        return taskRepo.findByProjectIdOrderBySortOrder(projectId).stream()
                .map(this::toDto)
                .toList();
    }

    public TaskDto create(Long projectId, CreateTaskRequest request) {
        var project = projectRepo.findById(projectId)
                .orElseThrow(() -> new ResourceNotFoundException("Project not found with id: " + projectId));

        Task task = new Task();
        task.setTitle(request.title());
        task.setDescription(request.description());
        task.setStatus(request.status() != null ? request.status() : TaskStatus.BACKLOG);
        task.setPriority(request.priority() != null ? request.priority() : TaskPriority.MEDIUM);
        task.setDueDate(request.dueDate());
        task.setSortOrder(request.sortOrder());
        task.setProject(project);

        return toDto(taskRepo.save(task));
    }

    public TaskDto updateStatus(Long taskId, UpdateTaskStatusRequest request) {
        Task task = taskRepo.findById(taskId)
                .orElseThrow(() -> new ResourceNotFoundException("Task not found with id: " + taskId));
        task.setStatus(request.status());
        return toDto(taskRepo.save(task));
    }

    public void delete(Long taskId) {
        if (!taskRepo.existsById(taskId)) {
            throw new ResourceNotFoundException("Task not found with id: " + taskId);
        }
        taskRepo.deleteById(taskId);
    }

    private TaskDto toDto(Task t) {
        return new TaskDto(
                t.getId(),
                t.getTitle(),
                t.getDescription(),
                t.getStatus(),
                t.getPriority(),
                t.getDueDate(),
                t.getSortOrder(),
                t.getProject().getId()
        );
    }
}

File: frontend/flowboard-ui/src/app/core/services/task.service.ts
import { Injectable, inject, signal } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { tap } from 'rxjs';

export type TaskStatus = 'BACKLOG' | 'TODO' | 'IN_PROGRESS' | 'DONE';
export type TaskPriority = 'LOW' | 'MEDIUM' | 'HIGH';

export interface Task {
  id: number;
  title: string;
  description: string | null;
  status: TaskStatus;
  priority: TaskPriority;
  dueDate: string | null;
  sortOrder: number;
  projectId: number;
}

export interface CreateTaskRequest {
  title: string;
  description?: string;
  status: TaskStatus;
  priority: TaskPriority;
  dueDate?: string;
  sortOrder?: number;
}

@Injectable({ providedIn: 'root' })
export class TaskService {
  private readonly http = inject(HttpClient);

  readonly tasks = signal<Task[]>([]);

  private url(projectId: number) {
    return `http://localhost:8080/api/projects/${projectId}/tasks`;
  }

  loadByProject(projectId: number) {
    return this.http.get<Task[]>(this.url(projectId)).pipe(
      tap(data => this.tasks.set(data))
    );
  }

  create(projectId: number, request: CreateTaskRequest) {
    return this.http.post<Task>(this.url(projectId), request).pipe(
      tap(task => this.tasks.update(list => [...list, task]))
    );
  }

  updateStatus(projectId: number, taskId: number, status: TaskStatus) {
    return this.http
      .patch<Task>(`${this.url(projectId)}/${taskId}/status`, { status })
      .pipe(
        tap(updated =>
          this.tasks.update(list =>
            list.map(t => (t.id === updated.id ? updated : t))
          )
        )
      );
  }

  delete(projectId: number, taskId: number) {
    return this.http.delete<void>(`${this.url(projectId)}/${taskId}`).pipe(
      tap(() => this.tasks.update(list => list.filter(t => t.id !== taskId)))
    );
  }
}
